### PR TITLE
[Circstore-149] Create API endpoint to retrieve patron IDs with expired sessions

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
Add API endpoint to retrieve all patrons Ids from patron_action_session table where last record's createdDate has expired according to configured timeout

Resolves story: https://issues.folio.org/browse/CIRCSTORE-149
## Approach
- update patron-action-session.raml to support retrieving patron IDs with expired sessions
- implement `getPatronActionSessionStorageExpiredSessionPatronIds `method in `PatronActionSessionAPI`
- create tests to `PatronActionSessionAPITest`
## Links
https://issues.folio.org/browse/CIRCSTORE-149

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -399,6 +399,10 @@
           "methods": ["DELETE"],
           "pathPattern": "/patron-action-session-storage/patron-action-sessions/{id}",
           "permissionsRequired": ["patron-action-session-storage.patron-action-sessions.item.delete"]
+        }, {
```
Comment (kostyantyn-kh) [inline descriptors/ModuleDescriptor-template.json]: Please update minor version

File: ramls/patron-action-session.raml [hunk 2]
```diff
@@ -12,6 +12,7 @@ types:
   patron-action-session: !include patron-action-session.json
   patron-action-sessions: !include patron-action-sessions.json
   errors: !include raml-util/schemas/errors.schema
+  patron-action-expired-ids-response: !include patron-action-expired-ids-response.json
```
Comment (kostyantyn-kh) [inline ramls/patron-action-session.raml]: Please update minor version

File: ramls/patron-action-session.raml [hunk 3]
```diff
@@ -47,3 +48,39 @@ resourceTypes:
...
         is: [validate]
+  /expired-session-patron-ids:
+      get:
+        queryParameters:
+          action_type:
+            description: Action type for patron notice session
+            type: string
+            enum: ["Check-in", "Check-out"]
+            required: true
+          last_time_action_limit:
```
Comment (kostyantyn-kh) [inline ramls/patron-action-session.raml]: Why a separate parameter was introduced? 
This value is pre-configured and is it it possible to retrieve value internally?

File: src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java [hunk 4]
```diff
@@ -55,6 +87,45 @@ public void deletePatronActionSessionStoragePatronActionSessionsByPatronSessionI
...
+  public void getPatronActionSessionStorageExpiredSessionPatronIds(
+    PatronActionSessionStorageExpiredSessionPatronIdsGetActionType actionType, String lastTimeActionLimit,
+    int limit, Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+
+    String tenantId = okapiHeaders.get(RestVerticle.OKAPI_HEADER_TENANT);
+    PgClientFutureAdapter pgClient = PgClientFutureAdapter.create(vertxContext, okapiHeaders);
+    DateTime dateTimeLimit;
+    try {
+      dateTimeLimit = DateTime.parse(lastTimeActionLimit);
+    }
```
Comment (kostyantyn-kh) [inline src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java]: Could this be at the  same line?
"} catch () {"

File: src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java [hunk 5]
```diff
@@ -55,6 +87,45 @@ public void deletePatronActionSessionStoragePatronActionSessionsByPatronSessionI
...
+    catch (Exception e){
```
Comment (kostyantyn-kh) [inline src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java]:
```suggestion
    catch (Exception e) {
```

File: ramls/patron-action-session.raml (hunk 3 above)
Comment (dmytrotur) [reply ramls/patron-action-session.raml]: Do you mean retrieve from somewhere like `mod-configuration`? If so, we shouldn't call other modules from `storage` modules, AFAIR.

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 6]
```diff
@@ -179,12 +185,99 @@ public void cannotUpdateNonExistentActionSession() throws InterruptedException,
...
+  }
+
+  @Test
+  public void cannotGetPatronActionSessionStorageExpiredSessionPatronIds()
+    throws InterruptedException, ExecutionException, TimeoutException, MalformedURLException {
+
+    checkGetExpiredSessionPatronIdsWithInvalidDate("Check-out", 10);
+  }
+
+  private JsonObject createSession(String actionType) {
```
Comment (kostyantyn-kh) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]:
```suggestion
  private JsonObject createActionSessionRecord(String actionType) {
```

File: src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java [hunk 7]
```diff
@@ -55,6 +87,45 @@ public void deletePatronActionSessionStoragePatronActionSessionsByPatronSessionI
...
+      .map(GetPatronActionSessionStorageExpiredSessionPatronIdsResponse::respond200WithApplicationJson)
+      .map(Response.class::cast)
+      .otherwise(this::mapExceptionToResponse)
+      .setHandler(asyncResultHandler);
+  }
+
+  private PatronActionExpiredIdsResponse mapPatronIdResponse(ResultSet resultSet) {
+    List<String> patronIds = resultSet.getRows()
+      .stream()
+      .map(json -> json.getString("patron_id"))
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java]: Why is there a JSON property which uses an underscore as a separator?

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 8]
```diff
@@ -179,12 +185,99 @@ public void cannotUpdateNonExistentActionSession() throws InterruptedException,
...
+    URL url = StorageTestSuite.storageUrl("/patron-action-session-storage/expired-session-patron-ids",
+      "action_type", actionType, "last_time_action_limit",
+      lastActionDateLimit.toString(ISODateTimeFormat.dateTime()), "limit", Integer.toString(limit));
+
+    JsonResponse response = performRequest(url);
+    assertThat(response, isOk());
+    return response.getJson().getJsonArray("patronIds").getList();
+  }
+
+  private void checkGetExpiredSessionPatronIdsWithInvalidDate(String actionType, int limit)
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: This appears to be only used once, why has it been extracted separate from that test?

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 9]
```diff
@@ -179,12 +185,99 @@ public void cannotUpdateNonExistentActionSession() throws InterruptedException,
...
+  private JsonObject createSession(String actionType) {
     JsonObject jsonObject = new JsonObject()
       .put("id", UUID.randomUUID().toString())
       .put("patronId", UUID.randomUUID().toString())
       .put("loanId", existingLoanId)
       .put("actionType", actionType);
     return jsonObject;
   }
+
+  private JsonObject createEntity(String patronId, String actionType, DateTime createdDate) {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]:
```suggestion
  private JsonObject createPatronAction(String patronId, String actionType, DateTime createdDate) {
```

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 10]
```diff
@@ -179,12 +185,99 @@ public void cannotUpdateNonExistentActionSession() throws InterruptedException,
...
+    throws InterruptedException, ExecutionException, TimeoutException, MalformedURLException {
+
+    URL url = StorageTestSuite.storageUrl("/patron-action-session-storage/expired-session-patron-ids",
+      "action_type", actionType, "last_time_action_limit",
+      "wrong date", "limit", Integer.toString(limit));
+    JsonResponse response = performRequest(url);
+    assertThat(response.getStatusCode(), is(422));
+  }
+
+  private JsonResponse performRequest(URL url)
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]:
```suggestion
  private JsonResponse get(URL url)
```

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 11]
```diff
@@ -179,12 +185,99 @@ public void cannotUpdateNonExistentActionSession() throws InterruptedException,
...
+
+    assertThat(getExpiredSessionPatronIds("Check-in", 10, DateTime.now().minusDays(2)).size(),
+      is(1));
+    assertThat(getExpiredSessionPatronIds("Check-out", 10, DateTime.now()).size(),
+      is(2));
+    assertThat(getExpiredSessionPatronIds("Check-in", 10, DateTime.now().minusDays(2)).get(0),
+      is(firstPatronId));
+    assertThat(getExpiredSessionPatronIds("Check-out", 10, DateTime.now().minusDays(2)).get(0),
+      is(firstPatronId));
+    assertThat(getExpiredSessionPatronIds("Check-out", 2, DateTime.now()).size(),
```
Comment (kostyantyn-kh) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: There is no difference between this and 2nd case
If it's intended to check "limit" parameter, it's better to pass limit=1

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (roman-barannyk) [reply descriptors/ModuleDescriptor-template.json]: Thank you for review, updated

File: ramls/patron-action-session.raml (hunk 2 above)
Comment (roman-barannyk) [reply ramls/patron-action-session.raml]: Updated

File: src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java (hunk 4 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java]: updated

File: src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java (hunk 5 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java]: fixed

File: src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java (hunk 7 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/PatronActionSessionAPI.java]: Thank you for review, updated

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java (hunk 8 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: updated

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java (hunk 9 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: renamed

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java (hunk 10 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: renamed

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java (hunk 11 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: Thanks, the test has been updated

File: ramls/patron-action-session.raml (hunk 3 above)
Comment (kostyantyn-kh) [reply ramls/patron-action-session.raml]: correct

Comment (kostyantyn-kh) [inline ramls/patron-action-session.raml]: It's better to name this parameter "period_of_inactivity" and make it integer
This value is stored in minutes in Settings - Circulation - Other settings

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 12]
```diff
@@ -173,18 +180,126 @@ public void canUpdatePatronActiveSession() throws InterruptedException, Malforme
...
+    createPatronActionSessionRecords(secondPatronId, "Check-out", DateTime.now().minus(1));
+
+    assertThat(getExpiredSessionPatronIds("Check-out", 10, DateTime.now()).size(), is(2));
+
+    JsonArray checkInJsonArray = getExpiredSessionPatronIds("Check-in", 10, DateTime.now().minusDays(2));
+    assertThat(checkInJsonArray.size(), is(1));
+    assertThat(checkInJsonArray.getJsonObject(0).getString("patronId"), is(firstPatronId));
+    assertThat(checkInJsonArray.getJsonObject(0).getString("actionType"), is("Check-in"));
+
+    JsonArray checkOutJsonArray = getExpiredSessionPatronIds("Check-out", 10, DateTime.now().minusDays(2));
```
Comment (kostyantyn-kh) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: Could you add test case to get expired sessions without and incorrect action type?

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 13]
```diff
@@ -173,18 +180,126 @@ public void canUpdatePatronActiveSession() throws InterruptedException, Malforme
...
+    assertThat(checkOutJsonArray.getJsonObject(0).getString("actionType"), is("Check-out"));
+
+    assertThat(getExpiredSessionPatronIds("Check-out", 1, DateTime.now()).size(), is(1));
+  }
+
+  @Test
+  public void canGetPatronActionSessionStorageExpiredSessionPatronIdsWithoutActionType()
+    throws InterruptedException, ExecutionException, TimeoutException, MalformedURLException {
+
+    String patronId = UUID.randomUUID().toString();
```
Comment (kostyantyn-kh) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: Could you enhance this test and add extra data for various action types, expiration date

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 14]
```diff
@@ -173,18 +180,126 @@ public void canUpdatePatronActiveSession() throws InterruptedException, Malforme
...
+
+    CompletableFuture<String> future = new CompletableFuture<>();
+    PostgresClient
+      .getInstance(StorageTestSuite.getVertx(), TENANT_ID)
+      .save(PATRON_ACTION_SESSION, session, update -> future.complete(update.result()));
+    future.join();
+    return session;
+  }
+
+  private JsonArray getExpiredSessionPatronIds(String actionType, int limit, DateTime lastActionDateLimit)
```
Comment (kostyantyn-kh) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: please remove code duplication, both _getExpiredSessionPartonIds_ methods have same body

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java (hunk 13 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: updated

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java (hunk 14 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]: Thanks, updated

File: src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java [hunk 15]
```diff
@@ -173,18 +180,144 @@ public void canUpdatePatronActiveSession() throws InterruptedException, Malforme
...
```
Comment (dmytrotur) [inline src/test/java/org/folio/rest/api/PatronActionSessionAPITest.java]:
```suggestion
  private JsonArray getExpiredPatronSessions(String actionType, int limit, DateTime lastActionDateLimit)
```

**Pull request merged**